langchain4j/src/main/java/dev/langchain4j/agent/tool/DefaultToolExecutor.java

```diff
@@ -109,14 +106,81 @@ static Object[] prepareArguments(
             if (argumentsMap.containsKey(parameterName)) {
                 Object argument = argumentsMap.get(parameterName);
                 Class<?> parameterType = parameters[i].getType();
-
-                arguments[i] = coerceArgument(argument, parameterName, parameterType);
+                if( parameterType == List.class || parameterType == Set.class || parameterType == Map.class)  {
+                    ParameterizedType parameterizedType = (ParameterizedType) parameters[i].getParameterizedType();
+                    arguments[i] = coerceCollectionArgument(argument, parameterName, parameterType, parameterizedType);
+                } if(parameterType.isArray() && argument instanceof Collection) {
+                    arguments[i] = coerceCollectionArgument(argument, parameterName, parameterType, parameterType.getComponentType());
+                }else {
+                    arguments[i] = coerceArgument(argument, parameterName, parameterType);
+                }
             }
         }
 
         return arguments;
     }
 
+    @SneakyThrows
+    static Object coerceCollectionArgument(
+            Object argument,
+            String parameterName,
+            Class<?> parameterType,
+            Object parameterizedType
+    )  {
+        String json = Json.toJson(argument);
+        if (parameterType == List.class) {
```

Comment:
4 blocks of logic in this method does not seem to be related, perhaprs it would be better to have 4 methods?